core/src/main/scala/kafka/log/Log.scala

```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
 
   def lastStableOffsetLag: Long = highWatermark - lastStableOffset
 
+  /**
+    * Fully materialize and return an offset snapshot including segment position info. This method will update
+    * the LogOffsetMetadata for the high watermark and log start offset if they are message-only. Throws an
+    * offset out of range error if the segment info cannot be loaded.
+    */
   def offsetSnapshot: LogOffsetSnapshot = {
+    val hw: LogOffsetMetadata = lock.synchronized {
```

Comment:
I think we can avoid the lock in the common case. Maybe we can rewrite this something like the following:
```scala
var hw = this.highWatermark
if (hw.messageOffsetOnly) {
  hw = maybeUpdateHighWatermarkMetadata()
}
```
where `maybeUpdateHighWatermarkMetadata` grabs the lock to safely update the value.